### PR TITLE
Spark archiver

### PR DESCRIPTION
**Pull request by YemreGurses**

### CODE REVIEW

File: tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      if (!notWrittenResources.isEmpty) {
+        import spark.implicits._
+        val notWrittenResourcesDs = spark.createDataset[FhirMappingResult](notWrittenResources)
+        this.writeDatasetToConfiguredFolder(mappingJobExecution, notWrittenResourcesDs, mappingUrl.get, FhirMappingErrorCodes.INVALID_RESOURCE)
+      }
+    }
+  }
+
+  /**
+   * Writes the dataset to the configured folder
```
Comment (sinaci) [inline tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala:53]: What is the configured folder here? I suppose there are subfolders w.r.t. the error type. If so, I recommend to describe this better in the javadoc.

File: tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      .csv(outputPath)
+
+    // Remove all files except the CSV file (to remove .crc files)
+    val srcFiles = FileUtil.listFiles(new File(outputPath))
+      .filterNot(f => f.getPath.endsWith(FileExtensions.CSV.toString))
+    srcFiles.foreach(f => f.delete())
+  }
+
+  /**
+   * Move file type data sources to archive folder
```
Comment (sinaci) [inline tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala:83]: I recommend to improve this description of the function since processArchiving is a braod term. It might be good to explain what is done in the match case conditions.

File: tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  type ArchiveModes = String
+  final val OFF = "off"
+  final val DELETE = "delete"
+  final val ARCHIVE = "archive"
+}
+
+/**
+ * Data source processing settings
+ * @param mappingErrorHandling Error handling type HALT or CONTINUE
+ * @param saveErroneousRecords If true, erroneous records will be saved to archive folder with the same path as the input file
```
Comment (sinaci) [inline tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala:16]: Do we have an archive folder which keeps the archived source files as well as the erroneous records? How do we differentiate them, by subfolders?

I see an error-folder-path configuration, is that under archive folder path config?

**Review by sinaci — CHANGES_REQUESTED** [review]

File: tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+          paths.foreach(relativePath => {
+            val relativeFilePath = Paths.get(dataFolderPath, relativePath)
+            this.processArchiveMode(mappingJobExecution, relativeFilePath)
+          })
+        })
+    }
+  }
+
+  /**
+   * Process archive mode according to the job settings
```
Comment (suatgonul) [inline tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala:144]: Can you please extend the description according to the archiving options available?

File: tofhir-engine/src/main/scala/io/tofhir/engine/mapping/FhirMappingJobManager.scala [hunk 5]
```diff
@@ -174,29 +174,29 @@ class FhirMappingJobManager(
...
-    val timeRange = getScheduledTimeRange(id, mappingJobScheduler.get.folderUri, startTime)
+    val timeRange = getScheduledTimeRange(job.id, mappingJobScheduler.get.folderUri, startTime)
     logger.info(s"Running scheduled job with the expression: ${schedulingSettings.cronExpression}")
     logger.info(s"Synchronizing data between ${timeRange._1} and ${timeRange._2}")
-    val mappingJobExecution = FhirMappingJobExecution(jobId = id, mappingTasks = tasks)
+    val mappingJobExecution = FhirMappingJobExecution(job = job, mappingTasks = tasks)
     executeMappingJob(mappingJobExecution, sourceSettings, sinkSettings, terminologyServiceSettings, identityServiceSettings, Some(timeRange))
       .map(_ => {
-        val writer = new FileWriter(s"${mappingJobScheduler.get.folderUri.getPath}/$id.txt", true)
+        val writer = new FileWriter(s"${mappingJobScheduler.get.folderUri.getPath}/$job.txt", true)
```
Comment (suatgonul) [inline tofhir-engine/src/main/scala/io/tofhir/engine/mapping/FhirMappingJobManager.scala:199]: Is this change intentional? Do we have test case covering this part?

File: tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package io.tofhir.engine.model
+
+import io.tofhir.engine.config.ErrorHandlingType.ErrorHandlingType
+import io.tofhir.engine.model.ArchiveModes.ArchiveModes
+
+object ArchiveModes extends Enumeration {
```
Comment (suatgonul) [inline tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala:6]: Can you please provide descriptions for these?

File: tofhir-server/src/main/scala/io/tofhir/server/service/ExecutionService.scala [hunk 7]
```diff
@@ -384,11 +396,32 @@ class ExecutionService(jobRepository: IJobRepository, mappingRepository: IMappin
...
           // Use the timestamp of the first log as the execution timestamp
           val timestamp = filteredLogs.select("@timestamp").first().getString(0)
           // Determine the status based on the success count
-          val status = if (successCount == 0) "FAILURE" else if (successCount != mappingUrls.length) "PARTIAL_SUCCESS" else "SUCCESS"
+          val results: Seq[String] = filteredLogs.select("result").collect().map(_.getString(0)).toSeq
+          var status = "STARTED"
+          if (!results.filterNot(_.contentEquals("STARTED")).isEmpty) {
+            // if there is a row with PARTIAL_SUCCESS, its status is PARTIAL_SUCCESS
+            val partialSuccessCount = results.count(_.contentEquals("PARTIAL_SUCCESS"))
+            if (partialSuccessCount > 0) {
```
Comment (suatgonul) [inline tofhir-server/src/main/scala/io/tofhir/server/service/ExecutionService.scala:407]: Duplicate?

File: tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala (hunk 3 above)
Comment (YemreGurses) [reply tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala:16]: We have two seperate configs: one for erroneous records and one for archiving. These configs specify the folder names for each and they work independently. Let's say these configs are set as following:
erroneousRecordsFolder: erroneous-records-folder
archiveFolder: archive-folder

The path structure of erroneous records:
`erroneous-records-folder/<error-type>/job-<jobId>/execution-<executionId>/<mappingUrl>/<random-generated-name-by-spark>.csv`

In the archiving mechanism, we use the same path of the file source relative the to repository (context) and put it in the configured archiving folder without changing the path. 
If path of csv file is this: `C:/dev/be/data-integration-suite/test-data/streaming-folder/patients/patients.csv` and context path is this: `C:/dev/be/data-integration-suite` 
We get the relative path of data source to the context path and process archiving according to it. 
Archiving path: `archive-folder/test-data/streaming-folder/patients/patients.csv`
That way, we can understand from which location the data source is archived.

File: tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala (hunk 6 above)
Comment (suatgonul) [reply tofhir-engine/src/main/scala/io/tofhir/engine/model/DataProcessingSettings.scala:9]: Can you please provide descriptions for different archive modes as well?

File: tofhir-engine/src/main/scala/io/tofhir/engine/mapping/FhirMappingJobManager.scala (hunk 5 above)
Comment (suatgonul) [reply tofhir-engine/src/main/scala/io/tofhir/engine/mapping/FhirMappingJobManager.scala]: @Okanmercan99 can you please add a test case covering this (to be implemented in another branch, resolving this issue for now to merge the main feature)

File: tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+   * @param df
+   * @param mappingJobExecution
+   */
+  private def applyArchiving(df: Dataset[FhirMappingResult], mappingJobExecution: FhirMappingJobExecution): Unit = {
+    val fileSystemSourceSettings = mappingJobExecution.job.sourceSettings.head._2.asInstanceOf[FileSystemSourceSettings]
+    // get data folder path from data source settings
+    val dataFolderPath = fileSystemSourceSettings.dataFolderPath
+    if (fileSystemSourceSettings.asStream) {
+      // get file names from df by grouping by path
+      val schema = schema_of_json(df.collect().head.source.get)
```
[the remark this answers is not in the document]
Comment (suatgonul) [reply tofhir-engine/src/main/scala/io/tofhir/engine/data/write/SparkArchiver.scala]: Resolving for now as we will be testing large datasets explicitly
